Added MetadataGranularities example for 3D Tiles 1.1

javagl opened the pull request:
Preview: https://github.com/javagl/3d-tiles-samples/tree/metadata-1.1/1.1/MetadataGranularities#metadata-granularities

(Note: This PR does not yet contain the addition of this sample to the table in the main `README.md`. There are several pending PRs that will likely already cause merge conflicts at this point. While they should usually be easy to resolve, it may be even easier if the tables are updated immediately before (or even after) merging the actual new samples)

lilleyse commented on 1.1/MetadataGranularities/tileset.json:
```diff
@@ -0,0 +1,362 @@
+    "class" : "exampleGroupMetadataClass",
+    "properties" : {
+      "color" : [ 64, 64, 255 ],
+      "priority" : "1"
```
Based on the class definition `priority` should be a number, not a string

lilleyse commented on 1.1/MetadataGranularities/tileset.json:
```diff
@@ -0,0 +1,362 @@
+        "class" : "exampleTileMetadataClass",
+        "properties" : {
+          "district" : "Callowhill",
+          "population" : "12"
```
Same comment here. `population` values should be numbers.

lilleyse commented on 1.1/MetadataGranularities/README.md:
```diff
@@ -0,0 +1,200 @@
+// to a certain position on the globe
+var tileset = viewer.scene.primitives.add(new Cesium.Cesium3DTileset({
+    url : 'http://localhost:8003/1.1/MetadataGranularities/tileset.json',
+    maximumScreenSpaceError: 1,
```
Use the default `maximumScreenSpaceError` if possible, which would mean modifying the geometric error in the tileset. I would be afraid of someone copying the code and using `maximumScreenSpaceError: 1` in their application without understanding the implications.

lilleyse commented on 1.1/MetadataGranularities/README.md:
```diff
@@ -0,0 +1,200 @@
+  labelEntity.label.text = "";
+  var picked = viewer.scene.pick(movement.endPosition);
+
+  //console.log("picked ", picked);
+  //debugger;
```
This looks like temporary debugging code that should be removed.

lilleyse commented on 1.1/MetadataGranularities/README.md:
```diff
@@ -0,0 +1,200 @@
+// is returned. 
+// Otherwise, 'undefined' is returned.
+var obtainTileMetadata = function(picked) {
+  if (!Cesium.defined(picked)) {
+    return undefined;
+  }
+  if (!Cesium.defined(picked.content)) {
+    return undefined;
+  }
+  if (!Cesium.defined(picked.content.tile)) {
+    return undefined;
+  }
+  if (!Cesium.defined(picked.content.tile.metadata)) {
+    return undefined;
+  }
```
You might be able to reduce the amount of code with [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining).

And speaking of using newer JS features, use let/const instead of var and run through prettier to fix any formatting inconsistencies.

javagl replied:
I think optional chaining is not supported... in sandcastles...? But will use `const`, that seems to work...

lilleyse replied:
https://github.com/CesiumGS/cesium/pull/10220 should fix the warning, just had to update some configs in sandcastle.

Also there's still one `var` in the code that should be a `let`.

lilleyse commented on 1.1/MetadataGranularities/README.md:
```diff
@@ -0,0 +1,200 @@
+    var propertyValue = metadata.getProperty(propertyKey);          
+    //console.log("propertyKey", propertyKey);
+    //console.log("propertyValue", propertyValue);
+    labelEntity.label.text += "  " + propertyKey + ": "+ propertyValue;
+    labelEntity.label.text += "\n";
```
[template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) will look nicer for a lot of these string builders

lilleyse commented on 1.1/MetadataGranularities/README.md:
```diff
@@ -0,0 +1,200 @@
+    return undefined;
+  }
+  var metadata = picked.content.tile.metadata;
+  var isTileMetadata = metadata instanceof Cesium.TileMetadata;
+  if (!isTileMetadata) {
+    return undefined;
+  }
```
When would it be a type other than `Cesium.TileMetadata`? Could the check simply be whether it's defined or not?

Also.. thinking about future demos.. what does this look like for implicit tiling? Is it the same API or different? CC @j9liu

j9liu replied:
`tile.metadata` can be of `Cesium.TileMetadata` (explicit) or `Cesium.ImplicitMetadataView`. However, the APIs for both are the same, so I think the check is not necessary. Along those lines, because the API for implicit tiling metadata should be the same, this sandcastle should function similarly for an implicit tileset.

javagl replied:
The `ImplicitMetadataView` was once called `ImplicitTileMetadata`, and I had figured that out in https://github.com/CesiumGS/3d-tiles-samples/blob/a546b82299ae5700525ba825564f857130a4e25f/EXT_mesh_features/CommonTestingSandcastleCode.txt#L119 , but didn't handle it here, because there are no implicit tilesets in this example. 

The general question of how rigurously all this should be checked remains open for me. E.g. I think that Peter once mentioned that ~"If _this_ is present, then _that_ is also present" (in view of the somewhat tedious `defined(...)` checks). But _omitting_ all of them seems dangerous for me....

lilleyse replied:
By removing the `instanceof` check (which I don't think is strictly necessary) and using optional chaining there's the potential to turn 100 lines of code to 10 lines of code:

**Before**
```js
// Given an object that was obtained via Scene#pick, examine it
// to see whether it contains tile metadata.
// If the object contains TileMetadata, then this metadata
// is returned.
// Otherwise, 'undefined' is returned.
const obtainTileMetadata = function (picked) {
  const metadata = picked?.content?.tile?.metadata;
  const isTileMetadata = metadata instanceof Cesium.TileMetadata;
  if (!isTileMetadata) {
    return undefined;
  }
  return metadata;
};

// Given an object that was obtained via Scene#pick, examine it
// to see whether it contains tileset metadata.
// If the object contains TilesetMetadata, then this metadata
// is returned.
// Otherwise, 'undefined' is returned.
const obtainTilesetMetadata = function (picked) {
  if (!Cesium.defined(picked)) {
    return undefined;
  }
  if (!Cesium.defined(picked.content)) {
    return undefined;
  }
  if (!Cesium.defined(picked.content.tileset)) {
    return undefined;
  }
  if (!Cesium.defined(picked.content.tileset.metadata)) {
    return undefined;
  }
  const metadata = picked.content.tileset.metadata.tileset;
  const isTilesetMetadata = metadata instanceof Cesium.TilesetMetadata;
  if (!isTilesetMetadata) {
    return undefined;
  }
  return metadata;
};

// Given an object that was obtained via Scene#pick, examine it
// to see whether it contains group metadata.
// If the object contains GroupMetadata, then this metadata
// is returned.
// Otherwise, 'undefined' is returned.
const obtainGroupMetadata = function (picked) {
  if (!Cesium.defined(picked)) {
    return undefined;
  }
  if (!Cesium.defined(picked.content)) {
    return undefined;
  }
  if (!Cesium.defined(picked.content.groupMetadata)) {
    return undefined;
  }
  const metadata = picked.content.groupMetadata;
  const isGroupMetadata = metadata instanceof Cesium.GroupMetadata;
  if (!isGroupMetadata) {
    return undefined;
  }
  return metadata;
};

// Given an object that was obtained via Scene#pick, examine it
// to see whether it contains content metadata.
// If the object contains ContentMetadata, then this metadata
// is returned.
// Otherwise, 'undefined' is returned.
const obtainContentMetadata = function (picked) {
  if (!Cesium.defined(picked)) {
    return undefined;
  }
  if (!Cesium.defined(picked.content)) {
    return undefined;
  }
  if (!Cesium.defined(picked.content.metadata)) {
    return undefined;
  }
  const metadata = picked.content.metadata;
  const isContentMetadata = metadata instanceof Cesium.ContentMetadata;
  if (!isContentMetadata) {
    return undefined;
  }
  return metadata;
};

// Install the handler that will check the element that is
// under the mouse cursor when the mouse is moved, and
// add any metadata that it contains to the label.
const handler = new Cesium.ScreenSpaceEventHandler(viewer.scene.canvas);
handler.setInputAction(function (movement) {
  labelEntity.label.text = "";
  const picked = viewer.scene.pick(movement.endPosition);

  const tilesetMetadata = obtainTilesetMetadata(picked);
  addMetadataToLabel("Tileset metadata", tilesetMetadata);

  const tileMetadata = obtainTileMetadata(picked);
  addMetadataToLabel("Tile metadata", tileMetadata);

  const groupMetadata = obtainGroupMetadata(picked);
  addMetadataToLabel("Group metadata", groupMetadata);

  const contentMetadata = obtainContentMetadata(picked);
  addMetadataToLabel("Content metadata", contentMetadata);

  const cartesian = viewer.scene.pickPosition(movement.endPosition);
  labelEntity.position = cartesian;
}, Cesium.ScreenSpaceEventType.MOUSE_MOVE);
```

**After**
```js
// Install the handler that will check the element that is
// under the mouse cursor when the mouse is moved, and
// add any metadata that it contains to the label.
const handler = new Cesium.ScreenSpaceEventHandler(viewer.scene.canvas);
handler.setInputAction(function (movement) {
  labelEntity.label.text = "";
  const picked = viewer.scene.pick(movement.endPosition);

  const tilesetMetadata = picked?.content?.tileset?.metadata?.tileset;
  const tileMetadata = picked?.content?.tile?.metadata;
  const groupMetadata = picked?.content?.groupMetadata;
  const contentMetadata = picked?.content?.metadata;
  
  addMetadataToLabel("Tileset metadata", tilesetMetadata);
  addMetadataToLabel("Tile metadata", tileMetadata);
  addMetadataToLabel("Group metadata", groupMetadata);
  addMetadataToLabel("Content metadata", contentMetadata);

  const cartesian = viewer.scene.pickPosition(movement.endPosition);
  labelEntity.position = cartesian;
}, Cesium.ScreenSpaceEventType.MOUSE_MOVE);
```

lilleyse replied:
This is also revealing some asymmetry in the API.

```js
  const tilesetMetadata = picked?.content?.tileset?.metadata?.tileset;
  const tileMetadata = picked?.content?.tile?.metadata;
  const groupMetadata = picked?.content?.groupMetadata;
  const contentMetadata = picked?.content?.metadata;
```

Instead I would expect the code below.

```js
  const tilesetMetadata = picked?.content?.tileset?.metadata;
  const tileMetadata = picked?.content?.tile?.metadata;
  const groupMetadata = picked?.content?.group?.metadata;
  const contentMetadata = picked?.content?.metadata;
```

@ptrgags @j9liu is this fixable?

ptrgags replied:
@lilleyse yes, I think this is pretty easy:

* for the tileset case, tha `tileset.metadata` object is currently a `Cesium3DTilesetMetadata` object which contains the schema, list of all group metadata objects, and the tileset metadata. If we make that object a private field of `Cesium3DTileset` and replace it with getters `tileset.metadata` (which will now return a `TilesetMetadata` object), `tileset.schema` and `tileset.groups`, these can just delegate to that object. This way, we don't have to clutter `Cesium3DTileset` with more responsibilities (it already does far too much)
* for the group metadata, we just need to add a simple wrapper object of a `Group` that has a `metadata` field for consistency of API.

ptrgags replied:
I opened https://github.com/CesiumGS/cesium/pull/10224 for this.

lilleyse replied:
And now merged. @javagl could you update to the new API?

lilleyse commented on 1.1/MetadataGranularities/README.md:
```diff
@@ -0,0 +1,200 @@
+# Metadata Granularities
+
+This sample demonstrates the use of [Metadata in 3D Tiles 1.1](https://github.com/CesiumGS/3d-tiles/blob/b7fa517e88149942bc6166b11944bac67565369e/specification/README.md#metadata) on different levels of granularity. The sample consists of a tileset with 4 tiles where each tile has 5 contents, and the contents are assigned to two different groups. Metadata is assigned to the tileset, the tiles, ea ...
```
Should this link point to the `draft-1.1` branch instead? We'll just need to update it when `draft-1.1` is merged.

lilleyse commented on 1.1/MetadataGranularities/README.md:
```diff
@@ -0,0 +1,200 @@
+  for (var i=0; i<propertyKeys.length; i++) {
+    var propertyKey = propertyKeys[i];
+    var propertyValue = metadata.getProperty(propertyKey);          
+    //console.log("propertyKey", propertyKey);
+    //console.log("propertyValue", propertyValue);
```
More commented code to remove

javagl commented:
Most of the inlined comments have been addressed. 

"Optional chaining" does not seem to work in Snadcastles (?).

---

Regarding the `obtain...` functions: The sandcastle also still carries a comment 
```
// (NOTE: The way how metadata is obtained may be refactored,
// see https://github.com/CesiumGS/cesium/issues/10015 )
``` 
pointing to https://github.com/CesiumGS/cesium/issues/10015

At the bottom of https://github.com/CesiumGS/3d-tiles/issues/641#issuecomment-1057359380 , I wondered...

> An aspect that is related to this discussion, but may be orthogonal, is: How to know _what_ a certain Metadata entity is associated with? One could pragmatically throw in some of the obvious options here, like defining an `enum OwnerType { TILESET, TILE, GROUP, CONTENT, ... }` and storing a `metadataEntity.ownerType = TILE`, but the devil is in the detail, and this may warrant some discussion in a dedicated issue at some point. 

and that might be relevant here. Speaking of API design... one (also pragmatic) generalization here could be such an `enum`, and then offer some function like
``` 
// Returns the metadata from the picked object, where "type" is 
// one of the OwnerType.TILESET/TILE/GROUP/CONTENT enum
// values
const obtainMetadata = function(picked, type) {
    ...
}
```
(basically as an _explicit_ version of [`getPropertyInherited`](https://github.com/CesiumGS/cesium/blob/9b93f47d43e95ec7dc610f7da45d4fc049a6ea90/Source/Scene/Cesium3DTileFeature.js#L263)), but this may require more thought.

lilleyse commented on 1.1/MetadataGranularities/README.md:
```diff
@@ -0,0 +1,198 @@
+// of metadata this is. The given metadata object
+// can be a TilesetMetadata, TileMetadata,
+// GroupMetadata, or ContentMetadata.
+const addMetadataToLabel = function (title, metadata) {
```
The more common style for local functions in CesiumJS is:

```js
function addMetadataToLabel(title, metadata) {
```

lilleyse commented:
I'm noticing a bit of label flickering. I think it happens when the cursor picks the label itself which causes `pickPosition` to be undefined. I'm not entirely sure why. Using a slightly larger pixel offset may help a little.

lilleyse commented:
The tileset geometric error is still a bit too small. The buildings disappear sooner than expected. You might have to really crank it up.

javagl commented:
I have replaced the `obtain...` methods with option chaining. Despite pulling the state that includes https://github.com/CesiumGS/cesium/pull/10220 , doing a `clean` and a `build`, it still shows the "Warning" indicators in the sandcastle code. It "works", but I wonder what might be necessary to remove the warnings. Maybe something's cached there, somewhere...

Also, I replaced the 3D Label with a 2D HTML overlay, to avoid the issue of accidentally picking the label during quick mouse movements.

lilleyse commented on 1.1/MetadataGranularities/README.md:
```diff
@@ -0,0 +1,109 @@
+  if (!Cesium.defined(propertyKeys)) {
+    return `(No properties for ${title})<br>`;
+  }
+  var html = `<b>${title}:</b><br>`;
```
Still two `var`'s left :grimacing:

lilleyse commented:
Alright I think that's it for my feedback. Just https://github.com/CesiumGS/3d-tiles-samples/pull/43#discussion_r833615539 and https://github.com/CesiumGS/3d-tiles-samples/pull/43#pullrequestreview-919294983. Then we can get this merged!

javagl commented:
> That's odd... it seems to work for me and warnings still get shown for legitimate problems.

There seems to have been a glitch with the update: Doing the "Fetch upstream" on GitHub apparently did not fetch the commits that had been done ~1 hour before. Now I fetched again, and it fetched 6 commits (including the API update), and now the warnings are gone.

The other points should also be addressed.
